qiskit/transpiler/passes/utils/gates_basis.py

```diff
@@ -18,22 +18,37 @@
 class GatesInBasis(AnalysisPass):
     """Check if all gates in a DAG are in a given set of gates"""
 
-    def __init__(self, basis_gates):
+    def __init__(self, basis_gates, target=None):
         """Initialize the GatesInBasis pass.
 
         Args:
             basis_gates (list): The list of strings representing the set of basis gates.
+            target (Target): If specified the target represnting the backend. If specified
+                this will be used instead of the ``basis_gates`` parameter
         """
         super().__init__()
         self._basis_gates = set(basis_gates)
+        self._target = target
 
     def run(self, dag):
         """Run the GatesInBasis pass on `dag`."""
         gates_out_of_basis = False
-        basic_instrs = {"measure", "reset", "barrier", "snapshot", "delay"}
-        for gate in dag._op_names:
-            if gate not in self._basis_gates and gate not in basic_instrs:
-                gates_out_of_basis = True
-                break
-
+        if self._target is not None:
+            qubit_map = {qubit: index for index, qubit in enumerate(dag.qubits)}
+            for gate in dag.topological_op_nodes():
+                # Barrier is universal and supported by all backends
+                if gate.name == "barrier":
+                    continue
+                if gate.name not in self._target:
+                    gates_out_of_basis = True
+                    break
+                if tuple(qubit_map[bit] for bit in gate.qargs) not in self._target[gate.name]:
+                    gates_out_of_basis = True
+                    break
+        else:
+            basic_instrs = {"measure", "reset", "barrier", "snapshot", "delay"}
```

Comment:
Something like this is more efficient. It doesn't really look harder to read to me, either (except that `union` is a method, rather than a function)
```python
            all_instrs = self._basis_gates.union({"measure", "reset", "barrier", "snapshot", "delay"})
            for gate in dag._op_names:
                if gate not in all_instrs:
```
I tested this with lists of random words and strings, i.e. find the number of things in list X that are in list Y or list Z.  Using the union is significantly faster when `len(X)` is relatively large, and is never slower. And the case that you have to iterate over all `dag._op_names` is common.

Maybe this belongs in a separate PR, just for purposes of finding bugs and performance regressions.

Reply:
I'll update it here. It is a bit unrelated but since we're basically rewriting the pass here I don't think there is any harm in making this improvement as part of this PR.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7548/commits/557362f6802939c244abc3470cd373ef6510510b